packages/combo-box/src/vaadin-combo-box-mixin.js

```diff
@@ -695,17 +695,20 @@ export const ComboBoxMixin = (subclass) =>
         }
       } else {
         const toLowerCase = (item) => item && item.toLowerCase && item.toLowerCase();
-        const itemsMatchedByLabel =
-          (this.filteredItems &&
-            this.filteredItems.filter(
-              (item) => toLowerCase(this._getItemLabel(item)) === toLowerCase(this._inputElementValue)
-            )) ||
-          [];
+
+        // Try to find an item whose label matches the input value. A matching item is searched from
+        // the filteredItems array (if available) and the selectedItem (if available).
+        const itemMatchingByLabel = [...(this.filteredItems || []), this.selectedItem].find((item) => {
```

Comment:
Cleaned up the logic a bit, but the main change here is that matching items are no longer searched only from the `filteredItems` array (which gets cleared by `comboBoxConnector`, producing the original bug) but also from `selectedItem`.